// @cloudflare/voprf-ts/lib/cjs/src/facade/impl/Codec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Codec = void 0;
const dleq_js_1 = require("../../dleq.js");
const oprf_js_1 = require("../../oprf.js");
class Codec {
    constructor(group, crypto) {
        this.group = group;
        this.crypto = crypto;
    }
    decodeElts(evaluated) {
        return evaluated.map(this.group.desElt.bind(this.group));
    }
    decodeScalars(evaluated) {
        return evaluated.map(this.group.desScalar.bind(this.group));
    }
    encodeArray(array) {
        return array.map((e) => e.serialize());
    }
    decodeProof(proof) {
        return dleq_js_1.DLEQProof.deserialize(this.group.id, proof, this.crypto);
    }
    encodeEvaluation(lib) {
        return {
            proof: lib.proof ? lib.proof.serialize() : undefined,
            mode: lib.mode,
            evaluated: this.encodeArray(lib.evaluated)
        };
    }
    encodeEvaluationRequest(evalRequest) {
        return {
            blinded: this.encodeArray(evalRequest.blinded)
        };
    }
    decodeEvaluation(fac) {
        return new oprf_js_1.Evaluation(fac.mode, this.decodeElts(fac.evaluated), fac.proof ? this.decodeProof(fac.proof) : undefined);
    }
    decodeEvaluationRequest(fac) {
        return new oprf_js_1.EvaluationRequest(this.decodeElts(fac.blinded));
    }
}
exports.Codec = Codec;
