// @cloudflare/voprf-ts/lib/cjs/src/facade/impl/KeyManagerImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyManagerImpl = void 0;
const OprfBaseImpl_js_1 = require("./OprfBaseImpl.js");
const keys_js_1 = require("../../keys.js");
class KeyManagerImpl extends OprfBaseImpl_js_1.OprfBaseImpl {
    sizes() {
        const internal = (0, keys_js_1.getKeySizes)(this.suite, this.crypto);
        return { publicKey: internal.Npk, privateKey: internal.Nsk };
    }
    validatePrivate(privateKey) {
        return (0, keys_js_1.validatePrivateKey)(this.suite, privateKey, this.crypto);
    }
    validatePublic(publicKey) {
        return (0, keys_js_1.validatePublicKey)(this.suite, publicKey, this.crypto);
    }
    randomPrivate() {
        return (0, keys_js_1.randomPrivateKey)(this.suite, this.crypto);
    }
    derivePrivate(seed, info) {
        return (0, keys_js_1.derivePrivateKey)(this.mode, this.suite, seed, info, this.crypto);
    }
    generatePublic(privateKey) {
        return (0, keys_js_1.generatePublicKey)(this.suite, privateKey, this.crypto);
    }
    generatePair() {
        return (0, keys_js_1.generateKeyPair)(this.suite, this.crypto);
    }
    derivePair(seed, info) {
        return (0, keys_js_1.deriveKeyPair)(this.mode, this.suite, seed, info, this.crypto);
    }
}
exports.KeyManagerImpl = KeyManagerImpl;
