// @cloudflare/voprf-ts/lib/cjs/src/facade/impl/ModeImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModeImpl = void 0;
const ServerImpl_js_1 = require("./ServerImpl.js");
const ClientImpl_js_1 = require("./ClientImpl.js");
const KeyManagerImpl_js_1 = require("./KeyManagerImpl.js");
const oprf_js_1 = require("../../oprf.js");
class ModeImpl {
    constructor(mode, suite, group, crypto) {
        this.mode = mode;
        this.suite = suite;
        this.group = group;
        this.crypto = crypto;
        this.keys = new KeyManagerImpl_js_1.KeyManagerImpl(...this.getBaseArgs());
        this.params = this.getParams();
    }
    getParams() {
        const [suite, group, hash, size] = (0, oprf_js_1.getOprfParams)(this.suite);
        const scalar = this.group.scalarSize();
        const elt = this.group.eltSize();
        return {
            mode: this.mode,
            suite,
            group,
            hash,
            sizes: {
                elt: elt,
                output: size,
                proof: scalar * 2,
                scalar
            }
        };
    }
    makeServer(privateKey) {
        return new ServerImpl_js_1.ServerImpl(privateKey, ...this.getBaseArgs());
    }
    makeClient(publicKey) {
        return new ClientImpl_js_1.ClientImpl(publicKey, ...this.getBaseArgs());
    }
    getBaseArgs() {
        return [this.mode, this.suite, this.crypto, this.group];
    }
}
exports.ModeImpl = ModeImpl;
