// @cloudflare/voprf-ts/lib/esm/src/facade/impl/Codec.js
import { DLEQProof } from '../../dleq.js';
import { Evaluation, EvaluationRequest } from '../../oprf.js';
export class Codec {
    constructor(group, crypto) {
        this.group = group;
        this.crypto = crypto;
    }
    decodeElts(evaluated) {
        return evaluated.map(this.group.desElt.bind(this.group));
    }
    decodeScalars(evaluated) {
        return evaluated.map(this.group.desScalar.bind(this.group));
    }
    encodeArray(array) {
        return array.map((e) => e.serialize());
    }
    decodeProof(proof) {
        return DLEQProof.deserialize(this.group.id, proof, this.crypto);
    }
    encodeEvaluation(lib) {
        return {
            proof: lib.proof ? lib.proof.serialize() : undefined,
            mode: lib.mode,
            evaluated: this.encodeArray(lib.evaluated)
        };
    }
    encodeEvaluationRequest(evalRequest) {
        return {
            blinded: this.encodeArray(evalRequest.blinded)
        };
    }
    decodeEvaluation(fac) {
        return new Evaluation(fac.mode, this.decodeElts(fac.evaluated), fac.proof ? this.decodeProof(fac.proof) : undefined);
    }
    decodeEvaluationRequest(fac) {
        return new EvaluationRequest(this.decodeElts(fac.blinded));
    }
}
